sphinx/testing/util.py

```diff
@@ -73,28 +77,75 @@ def etree_parse(path: str) -> Any:
 
 
 class SphinxTestApp(sphinx.application.Sphinx):
-    """
-    A subclass of :class:`Sphinx` that runs on the test root, with some
-    better default values for the initialization parameters.
+    """A subclass of :class:`~sphinx.application.Sphinx` for tests.
+
+    The constructor uses some better default values for the initialization
+    parameters and supports arbitrary keywords stored in the :attr:`extras`
+    read-only mapping.
+
+    It is recommended to use::
+
+        @pytest.mark.sphinx('html')
+        def test(app):
+            app = ...
+
+    instead of::
+
+        def test():
+            app = SphinxTestApp('html', srcdir=srcdir)
+
+    In the former case, the 'app' fixture takes care of setting the source
+    directory, whereas in the latter, the user must provide it themselves.
     """
 
-    _status: StringIO
-    _warning: StringIO
+    # Allow the builder name to be passed as a keyword argument
+    # but only make it positional-only for ``pytest.mark.sphinx``
+    # so that an exception can be raised if the constructor is
+    # directly called and multiple values for the builder name
+    # are given.
 
     def __init__(
         self,
+        /,
         buildername: str = 'html',
-        srcdir: Path | None = None,
-        builddir: Path | None = None,
+        *,
+        srcdir: Path,
+        confoverrides: dict[str, Any] | None = None,
+        status: StringIO | None = None,
+        warning: StringIO | None = None,
         freshenv: bool = False,
-        confoverrides: dict | None = None,
-        status: IO | None = None,
-        warning: IO | None = None,
+        warningiserror: bool = False,
         tags: list[str] | None = None,
-        docutils_conf: str | None = None,
+        verbosity: int = 0,
         parallel: int = 0,
+        keep_going: bool = False,
+        # extra constructor arguments
+        builddir: Path | None = None,
+        docutils_conf: str | None = None,
+        # unknown keyword arguments
+        **extras: Any,
     ) -> None:
```

Comment:
If I were to play the Devil's advocate I'd say "this was never documented and was still marked as experimental since 1.6". I can find a way to make it find without breaking changes though. The issue is that I honestly don't kwow who would actually pass the arguments without using keywords. 

Having something like that would force people to understand which parameter they are using. Note that the class inherits from Sphinx but did not even expose every other constructor arguments that the super class did (like verbosity, which I think *should* be exposed because the marker is really many to be used as a "constructor configuration" IMO). 

But let's avoid breaking changes, and I'll just parse the things differently.

Reply:
- The '/' is not really an issue (it just allows you that 'self' as an extra keyword).
- The '*' is needed (see explanation below)